### DISCUSSION TITLE
Stateless BTRFS with systemd-nspawn as opposed to AB partitions

### DISCUSSION THREAD

**Discussion by NjlsShade** [linuxdeepin/developer-center#3176] [Features Request & Ideas | 特性请求 & 头脑风暴]
As opposed to using the current AB root partition schema, which reserves space and is difficult to modify and adapt to future needs, I believe using BtrFS subvolumes can replace this functionality entirely.

A "**roota**" and "**data**" subvolume can be created, then a snapshot of the root subvolume called “**rootb**” can be used in place of backing up to a second root partition during updates. You can either delete the existing rootb snapshot to make a new one before updating, or you can simply apply the update directly to rootb, then passively switch to the rootb subvolume as the active root on next reboot.

> Root: /@​roota ro /
> Root: /@​rootb
> Data: /@​data rw /data

To ensure system stability the root subvolume should be read-only (with the option to switch to rw for advanced users), and a per-user container subvolume can be created under the data subvolume.

> Containers: /@​data/containers/$user/@​debian-sid

Here we can deploy a **systemd-nspawn** container of Debian Sid, in the container we have a user automatically created with the same name and UID as the host user (systemd-nspawn also allows bind mounting, so we'll bind the user's home directly for file access, themeing and configurations) any application the user installs would be installed within the container as opposed to the read-only root subvolume. The deepin launcher can be adjusted to look for and launch applications under the read-only root and the container, allowing applications like deepin browser to be included in regular system updates and be immutable.

This setup allows you to stay on a stable base like Debian 10 while still allowing the use of up-to-date applications. I currently use systemd-nspawn on deepin to run the latest version of Flatpak, as many flatpak applications I use require version 1.12+, only available on the Debian Sid branch.

Just a segestion, I’m not sure if you already have plans on moving to a stateless design in some other way, however, I just wanted to throw this option out there as I think it’s a very friendly setup.

**Comment by justforlxz** [comment]
Great idea, I'm looking into replacing the AB partition with a btrfs filesystem, but I didn't think of using systemd-nspawn for less work, it's a good idea.

**Reply by NjlsShade** [reply]
Thanks! I think systemd-nspawn should provide a capable namespace environment if you decide to go that route, It integrates really well and is easy to manage, especially when paired with btrfs.

Among other things, I use it for gaming via steam (saves me from getting my system all messy with all those 32bit libraries, lol), as well as providing a custom driver for my keyboard and mouse via ckb-next (which I'd have to pull in dev tools to compile manually on Debian 10). It seems to interface with hardware just fine when configured correctly.

Would be really cool if some form of namespace or other container option gets adopted, as I love the idea of keeping the OS both stable and compatible.
